// igniteui-angular/esm5/lib/grids/selection/selection.service.js
import { __assign, __decorate, __read, __spread } from "tslib";
import { Injectable, EventEmitter, NgZone } from '@angular/core';
import { FilteringExpressionsTree } from '../../data-operations/filtering-expressions-tree';
import { isEdge } from '../../core/utils';
// TODO: Refactor - export in a separate file
var IgxRow = /** @class */ (function () {
    function IgxRow(id, index, data) {
        this.id = id;
        this.index = index;
        this.data = data;
    }
    IgxRow.prototype.createEditEventArgs = function () {
        return {
            rowID: this.id,
            oldValue: __assign({}, this.data),
            newValue: this.newData,
            cancel: false
        };
    };
    return IgxRow;
}());
export { IgxRow };
var IgxCell = /** @class */ (function () {
    function IgxCell(id, rowIndex, column, value, editValue, rowData) {
        this.id = id;
        this.rowIndex = rowIndex;
        this.column = column;
        this.value = value;
        this.editValue = editValue;
        this.rowData = rowData;
    }
    IgxCell.prototype.castToNumber = function (value) {
        if (this.column.dataType === 'number' && !this.column.inlineEditorTemplate) {
            var v = parseFloat(value);
            return !isNaN(v) && isFinite(v) ? v : 0;
        }
        return value;
    };
    IgxCell.prototype.createEditEventArgs = function () {
        return {
            rowID: this.id.rowID,
            cellID: this.id,
            oldValue: this.value,
            newValue: this.editValue,
            cancel: false
        };
    };
    return IgxCell;
}());
export { IgxCell };
var IgxGridCRUDService = /** @class */ (function () {
    function IgxGridCRUDService() {
        this.cell = null;
        this.row = null;
        this.isInCompositionMode = false;
    }
    IgxGridCRUDService.prototype.createCell = function (cell) {
        return new IgxCell(cell.cellID, cell.rowIndex, cell.column, cell.value, cell.value, cell.row.rowData);
    };
    IgxGridCRUDService.prototype.createRow = function (cell) {
        return new IgxRow(cell.id.rowID, cell.rowIndex, cell.rowData);
    };
    IgxGridCRUDService.prototype.sameRow = function (rowID) {
        return this.row && this.row.id === rowID;
    };
    IgxGridCRUDService.prototype.sameCell = function (cell) {
        return (this.cell.id.rowID === cell.id.rowID &&
            this.cell.id.columnID === cell.id.columnID);
    };
    Object.defineProperty(IgxGridCRUDService.prototype, "inEditMode", {
        get: function () {
            return !!this.cell;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridCRUDService.prototype, "rowEditing", {
        get: function () {
            return this.grid.rowEditable;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridCRUDService.prototype, "primaryKey", {
        get: function () {
            return this.grid.primaryKey;
        },
        enumerable: true,
        configurable: true
    });
    IgxGridCRUDService.prototype.beginRowEdit = function () {
        if (this.grid.rowEditable && (this.grid.primaryKey === undefined || this.grid.primaryKey === null)) {
            console.warn('The grid must have a `primaryKey` specified when using `rowEditable`!');
        }
        this.row = this.createRow(this.cell);
        var args = {
            rowID: this.row.id,
            oldValue: this.row.data,
            cancel: false
        };
        this.grid.onRowEditEnter.emit(args);
        if (args.cancel) {
            this.endRowEdit();
            return;
        }
        this.row.transactionState = this.grid.transactions.getAggregatedValue(this.row.id, true);
        this.grid.transactions.startPending();
        this.grid.openRowOverlay(this.row.id);
    };
    IgxGridCRUDService.prototype.endRowEdit = function () {
        this.row = null;
    };
    IgxGridCRUDService.prototype.begin = function (cell) {
        var newCell = this.createCell(cell);
        newCell.primaryKey = this.primaryKey;
        var args = {
            cellID: newCell.id,
            rowID: newCell.id.rowID,
            oldValue: newCell.value,
            cancel: false
        };
        this.grid.onCellEditEnter.emit(args);
        if (args.cancel) {
            this.end();
            return;
        }
        if (this.rowEditing) {
            if (this.row && !this.sameRow(newCell.id.rowID)) {
                this.grid.endEdit(true);
                this.cell = newCell;
                this.beginRowEdit();
                return;
            }
            this.cell = newCell;
            if (!this.row) {
                this.beginRowEdit();
                return;
            }
        }
        else {
            this.cell = newCell;
            this.endRowEdit();
        }
    };
    IgxGridCRUDService.prototype.end = function () {
        this.cell = null;
    };
    IgxGridCRUDService.prototype.enterEditMode = function (cell) {
        if (this.isInCompositionMode) {
            return;
        }
        if (cell && cell.column.editable && !cell.row.deleted) {
            if (this.inEditMode) {
                this.grid.endEdit(true);
                this.grid.tbody.nativeElement.focus();
            }
            else {
                this.begin(cell);
            }
        }
    };
    IgxGridCRUDService.prototype.exitEditMode = function () {
        if (this.isInCompositionMode) {
            return;
        }
        if (this.inEditMode) {
            var args = this.cell.createEditEventArgs();
            this.grid.onCellEditCancel.emit(args);
            if (args.cancel) {
                return;
            }
            this.grid.endEdit(false);
            if (isEdge()) {
                this.grid.cdr.detectChanges();
            }
            this.grid.tbody.nativeElement.focus();
        }
    };
    IgxGridCRUDService.prototype.isInEditMode = function (rowIndex, columnIndex) {
        if (!this.cell) {
            return false;
        }
        return this.cell.column.index === columnIndex && this.cell.rowIndex === rowIndex;
    };
    IgxGridCRUDService = __decorate([
        Injectable()
    ], IgxGridCRUDService);
    return IgxGridCRUDService;
}());
export { IgxGridCRUDService };
var IgxGridSelectionService = /** @class */ (function () {
    function IgxGridSelectionService(zone) {
        this.zone = zone;
        this.dragMode = false;
        this.keyboardState = {};
        this.pointerState = {};
        this.columnsState = {};
        this.selection = new Map();
        this.temp = new Map();
        this._ranges = new Set();
        this.rowSelection = new Set();
        this.columnSelection = new Set();
        this.initPointerState();
        this.initKeyboardState();
        this.initColumnsState();
    }
    Object.defineProperty(IgxGridSelectionService.prototype, "ranges", {
        /**
         * Returns the current selected ranges in the grid from both
         * keyboard and pointer interactions
         */
        get: function () {
            // The last action was keyboard + shift selection -> add it
            this.addKeyboardRange();
            var ranges = Array.from(this._ranges).map(function (range) { return JSON.parse(range); });
            // No ranges but we have a focused cell -> add it
            if (!ranges.length && this.activeElement && this.grid.isCellSelectable) {
                ranges.push(this.generateRange(this.activeElement));
            }
            return ranges;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridSelectionService.prototype, "primaryButton", {
        get: function () {
            return this.pointerState.primaryButton;
        },
        set: function (value) {
            this.pointerState.primaryButton = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Resets the keyboard state
     */
    IgxGridSelectionService.prototype.initKeyboardState = function () {
        this.keyboardState.node = null;
        this.keyboardState.shift = false;
        this.keyboardState.range = null;
        this.keyboardState.active = false;
    };
    /**
     * Resets the pointer state
     */
    IgxGridSelectionService.prototype.initPointerState = function () {
        this.pointerState.node = null;
        this.pointerState.ctrl = false;
        this.pointerState.shift = false;
        this.pointerState.range = null;
        this.pointerState.primaryButton = true;
    };
    /**
     * Resets the columns state
     */
    IgxGridSelectionService.prototype.initColumnsState = function () {
        this.columnsState.field = null;
        this.columnsState.range = [];
    };
    /**
     * Adds a single node.
     * Single clicks | Ctrl + single clicks on cells is the usual case.
     */
    IgxGridSelectionService.prototype.add = function (node, addToRange) {
        if (addToRange === void 0) { addToRange = true; }
        this.selection.has(node.row) ? this.selection.get(node.row).add(node.column) :
            this.selection.set(node.row, new Set()).get(node.row).add(node.column);
        if (addToRange) {
            this._ranges.add(JSON.stringify(this.generateRange(node)));
        }
    };
    /**
     * Adds the active keyboard range selection (if any) to the `ranges` meta.
     */
    IgxGridSelectionService.prototype.addKeyboardRange = function () {
        if (this.keyboardState.range) {
            this._ranges.add(JSON.stringify(this.keyboardState.range));
        }
    };
    IgxGridSelectionService.prototype.remove = function (node) {
        if (this.selection.has(node.row)) {
            this.selection.get(node.row).delete(node.column);
        }
        if (this.isActiveNode(node)) {
            this.activeElement = null;
        }
        this._ranges.delete(JSON.stringify(this.generateRange(node)));
    };
    IgxGridSelectionService.prototype.isInMap = function (node) {
        return (this.selection.has(node.row) && this.selection.get(node.row).has(node.column)) ||
            (this.temp.has(node.row) && this.temp.get(node.row).has(node.column));
    };
    IgxGridSelectionService.prototype.selected = function (node) {
        return (this.isActiveNode(node) && this.grid.isCellSelectable) || this.isInMap(node);
    };
    IgxGridSelectionService.prototype.isActiveNode = function (node) {
        if (this.activeElement) {
            var isActive = this.activeElement.column === node.column && this.activeElement.row === node.row;
            if (this.grid.hasColumnLayouts) {
                var layout = this.activeElement.layout;
                return isActive && this.isActiveLayout(layout, node.layout);
            }
            return isActive;
        }
        return false;
    };
    IgxGridSelectionService.prototype.isActiveLayout = function (current, target) {
        return current.columnVisibleIndex === target.columnVisibleIndex;
    };
    IgxGridSelectionService.prototype.addRangeMeta = function (node, state) {
        this._ranges.add(JSON.stringify(this.generateRange(node, state)));
    };
    IgxGridSelectionService.prototype.removeRangeMeta = function (node, state) {
        this._ranges.delete(JSON.stringify(this.generateRange(node, state)));
    };
    /**
     * Generates a new selection range from the given `node`.
     * If `state` is passed instead it will generate the range based on the passed `node`
     * and the start node of the `state`.
     */
    IgxGridSelectionService.prototype.generateRange = function (node, state) {
        if (!state) {
            return {
                rowStart: node.row,
                rowEnd: node.row,
                columnStart: node.column,
                columnEnd: node.column
            };
        }
        var _a = state.node, row = _a.row, column = _a.column;
        var rowStart = Math.min(node.row, row);
        var rowEnd = Math.max(node.row, row);
        var columnStart = Math.min(node.column, column);
        var columnEnd = Math.max(node.column, column);
        return { rowStart: rowStart, rowEnd: rowEnd, columnStart: columnStart, columnEnd: columnEnd };
    };
    /**
     *
     */
    IgxGridSelectionService.prototype.keyboardStateOnKeydown = function (node, shift, shiftTab) {
        this.keyboardState.active = true;
        this.initPointerState();
        this.keyboardState.shift = shift && !shiftTab;
        if (!this.grid.navigation.isDataRow(node.row)) {
            return;
        }
        // Kb navigation with shift and no previous node.
        // Clear the current selection init the start node.
        if (this.keyboardState.shift && !this.keyboardState.node) {
            this.clear();
            this.keyboardState.node = Object.assign({}, node);
        }
    };
    IgxGridSelectionService.prototype.keyboardStateOnFocus = function (node, emitter, dom) {
        var kbState = this.keyboardState;
        // Focus triggered by keyboard navigation
        if (kbState.active) {
            if (isChromium()) {
                this._moveSelectionChrome(dom);
            }
            // Start generating a range if shift is hold
            if (kbState.shift) {
                this.dragSelect(node, kbState);
                kbState.range = this.generateRange(node, kbState);
                emitter.emit(this.generateRange(node, kbState));
                return;
            }
            this.initKeyboardState();
            this.clear();
            this.add(node);
        }
    };
    IgxGridSelectionService.prototype.pointerDown = function (node, shift, ctrl) {
        this.addKeyboardRange();
        this.initKeyboardState();
        this.pointerState.ctrl = ctrl;
        this.pointerState.shift = shift;
        // No ctrl key pressed - no multiple selection
        if (!ctrl) {
            this.clear();
        }
        if (shift) {
            // No previously 'clicked' node. Use the last active node.
            if (!this.pointerState.node) {
                this.pointerState.node = this.activeElement || node;
            }
            this.pointerDownShiftKey(node);
            this.clearTextSelection();
            return;
        }
        this.removeRangeMeta(node);
        this.pointerState.node = node;
    };
    IgxGridSelectionService.prototype.pointerDownShiftKey = function (node) {
        this.clear();
        this.selectRange(node, this.pointerState);
    };
    IgxGridSelectionService.prototype.mergeMap = function (target, source) {
        var iterator = source.entries();
        var pair = iterator.next();
        var key;
        var value;
        var _loop_1 = function () {
            var _a;
            _a = __read(pair.value, 2), key = _a[0], value = _a[1];
            if (target.has(key)) {
                var newValue_1 = target.get(key);
                value.forEach(function (record) { return newValue_1.add(record); });
                target.set(key, newValue_1);
            }
            else {
                target.set(key, value);
            }
            pair = iterator.next();
        };
        while (!pair.done) {
            _loop_1();
        }
    };
    IgxGridSelectionService.prototype.pointerEnter = function (node, event) {
        // https://www.w3.org/TR/pointerevents/#the-button-property
        this.dragMode = event.buttons === 1 && event.button === -1;
        if (!this.dragMode) {
            return false;
        }
        this.clearTextSelection();
        // If the users triggers a drag-like event by first clicking outside the grid cells
        // and then enters in the grid body we may not have a initial pointer starting node.
        // Assume the first pointerenter node is where we start.
        if (!this.pointerState.node) {
            this.pointerState.node = node;
        }
        this.pointerState.ctrl ? this.selectRange(node, this.pointerState, this.temp) :
            this.dragSelect(node, this.pointerState);
        return true;
    };
    IgxGridSelectionService.prototype.pointerUp = function (node, emitter) {
        var _this = this;
        if (this.dragMode) {
            this.restoreTextSelection();
            this.addRangeMeta(node, this.pointerState);
            this.mergeMap(this.selection, this.temp);
            this.zone.runTask(function () { return emitter.emit(_this.generateRange(node, _this.pointerState)); });
            this.temp.clear();
            this.dragMode = false;
            return true;
        }
        if (this.pointerState.shift) {
            this.clearTextSelection();
            this.restoreTextSelection();
            this.addRangeMeta(node, this.pointerState);
            emitter.emit(this.generateRange(node, this.pointerState));
            return true;
        }
        this.add(node);
        return false;
    };
    IgxGridSelectionService.prototype.selectRange = function (node, state, collection) {
        if (collection === void 0) { collection = this.selection; }
        if (collection === this.temp) {
            collection.clear();
        }
        var _a = this.generateRange(node, state), rowStart = _a.rowStart, rowEnd = _a.rowEnd, columnStart = _a.columnStart, columnEnd = _a.columnEnd;
        for (var i = rowStart; i <= rowEnd; i++) {
            for (var j = columnStart; j <= columnEnd; j++) {
                collection.has(i) ? collection.get(i).add(j) :
                    collection.set(i, new Set()).get(i).add(j);
            }
        }
    };
    IgxGridSelectionService.prototype.dragSelect = function (node, state) {
        if (!this.pointerState.ctrl) {
            this.selection.clear();
        }
        this.selectRange(node, state);
    };
    IgxGridSelectionService.prototype.clear = function (clearAcriveEl) {
        if (clearAcriveEl === void 0) { clearAcriveEl = false; }
        if (clearAcriveEl) {
            this.activeElement = null;
        }
        this.selection.clear();
        this.temp.clear();
        this._ranges.clear();
    };
    IgxGridSelectionService.prototype.clearTextSelection = function () {
        var selection = window.getSelection();
        if (selection.rangeCount) {
            this._selectionRange = selection.getRangeAt(0);
            this._selectionRange.collapse(true);
            selection.removeAllRanges();
        }
    };
    IgxGridSelectionService.prototype.restoreTextSelection = function () {
        var selection = window.getSelection();
        if (!selection.rangeCount) {
            selection.addRange(this._selectionRange || document.createRange());
        }
    };
    /**
     * (╯°□°）╯︵ ┻━┻
     * Chrome and Chromium don't care about the active
     * range after keyboard navigation, thus this.
     */
    IgxGridSelectionService.prototype._moveSelectionChrome = function (node) {
        var selection = window.getSelection();
        selection.removeAllRanges();
        var range = new Range();
        range.selectNode(node);
        range.collapse(true);
        selection.addRange(range);
    };
    /** Returns array of the selected row id's. */
    IgxGridSelectionService.prototype.getSelectedRows = function () {
        return this.rowSelection.size ? Array.from(this.rowSelection.keys()) : [];
    };
    /** Clears row selection, if filtering is applied clears only selected rows from filtered data. */
    IgxGridSelectionService.prototype.clearRowSelection = function (event) {
        var _this = this;
        var removedRec = this.isFilteringApplied() ?
            this.getRowIDs(this.allData).filter(function (rID) { return _this.isRowSelected(rID); }) : this.getSelectedRows();
        var newSelection = this.isFilteringApplied() ? this.getSelectedRows().filter(function (x) { return !removedRec.includes(x); }) : [];
        this.emitRowSelectionEvent(newSelection, [], removedRec, event);
    };
    /** Select all rows, if filtering is applied select only from filtered data. */
    IgxGridSelectionService.prototype.selectAllRows = function (event) {
        var _this = this;
        var allRowIDs = this.getRowIDs(this.allData);
        var addedRows = allRowIDs.filter(function (rID) { return !_this.isRowSelected(rID); });
        var newSelection = this.rowSelection.size ? this.getSelectedRows().concat(addedRows) : addedRows;
        this.emitRowSelectionEvent(newSelection, addedRows, [], event);
    };
    /** Select the specified row and emit event. */
    IgxGridSelectionService.prototype.selectRowById = function (rowID, clearPrevSelection, event) {
        if (!this.grid.isRowSelectable || this.isRowDeleted(rowID)) {
            return;
        }
        clearPrevSelection = !this.grid.isMultiRowSelectionEnabled || clearPrevSelection;
        var newSelection = clearPrevSelection ? [rowID] : this.getSelectedRows().indexOf(rowID) !== -1 ?
            this.getSelectedRows() : __spread(this.getSelectedRows(), [rowID]);
        var removed = clearPrevSelection ? this.getSelectedRows() : [];
        this.emitRowSelectionEvent(newSelection, [rowID], removed, event);
    };
    /** Deselect the specified row and emit event. */
    IgxGridSelectionService.prototype.deselectRow = function (rowID, event) {
        if (!this.isRowSelected(rowID)) {
            return;
        }
        var newSelection = this.getSelectedRows().filter(function (r) { return r !== rowID; });
        if (this.rowSelection.size && this.rowSelection.has(rowID)) {
            this.emitRowSelectionEvent(newSelection, [], [rowID], event);
        }
    };
    /** Select specified rows. No event is emitted. */
    IgxGridSelectionService.prototype.selectRowsWithNoEvent = function (rowIDs, clearPrevSelection) {
        var _this = this;
        if (clearPrevSelection) {
            this.rowSelection.clear();
        }
        rowIDs.forEach(function (rowID) { _this.rowSelection.add(rowID); });
        this.allRowsSelected = undefined;
    };
    /** Deselect specified rows. No event is emitted. */
    IgxGridSelectionService.prototype.deselectRowsWithNoEvent = function (rowIDs) {
        var _this = this;
        rowIDs.forEach(function (rowID) { return _this.rowSelection.delete(rowID); });
        this.allRowsSelected = undefined;
    };
    IgxGridSelectionService.prototype.isRowSelected = function (rowID) {
        return this.rowSelection.size > 0 && this.rowSelection.has(rowID);
    };
    /** Select range from last selected row to the current specified row. */
    IgxGridSelectionService.prototype.selectMultipleRows = function (rowID, rowData, event) {
        var _this = this;
        this.allRowsSelected = undefined;
        if (!this.rowSelection.size || this.isRowDeleted(rowID)) {
            this.selectRowById(rowID);
            return;
        }
        var gridData = this.allData;
        var lastRowID = this.getSelectedRows()[this.rowSelection.size - 1];
        var currIndex = gridData.indexOf(this.getRowDataById(lastRowID));
        var newIndex = gridData.indexOf(rowData);
        var rows = gridData.slice(Math.min(currIndex, newIndex), Math.max(currIndex, newIndex) + 1);
        var added = this.getRowIDs(rows).filter(function (rID) { return !_this.isRowSelected(rID); });
        var newSelection = this.getSelectedRows().concat(added);
        this.emitRowSelectionEvent(newSelection, added, [], event);
    };
    IgxGridSelectionService.prototype.areAllRowSelected = function () {
        if (!this.grid.data) {
            return false;
        }
        if (this.allRowsSelected !== undefined) {
            return this.allRowsSelected;
        }
        var dataItemsID = this.getRowIDs(this.allData);
        return this.allRowsSelected = Math.min(this.rowSelection.size, dataItemsID.length) > 0 &&
            new Set(Array.from(this.rowSelection.values()).concat(dataItemsID)).size === this.rowSelection.size;
    };
    IgxGridSelectionService.prototype.hasSomeRowSelected = function () {
        var _this = this;
        var filteredData = this.isFilteringApplied() ?
            this.getRowIDs(this.grid.filteredData).some(function (rID) { return _this.isRowSelected(rID); }) : true;
        return this.rowSelection.size > 0 && filteredData && !this.areAllRowSelected();
    };
    Object.defineProperty(IgxGridSelectionService.prototype, "filteredSelectedRowIds", {
        get: function () {
            var _this = this;
            return this.isFilteringApplied() ?
                this.getRowIDs(this.allData).filter(function (rowID) { return _this.isRowSelected(rowID); }) :
                this.getSelectedRows().filter(function (rowID) { return !_this.isRowDeleted(rowID); });
        },
        enumerable: true,
        configurable: true
    });
    IgxGridSelectionService.prototype.emitRowSelectionEvent = function (newSelection, added, removed, event) {
        var currSelection = this.getSelectedRows();
        if (this.areEqualCollections(currSelection, newSelection)) {
            return;
        }
        var args = {
            oldSelection: currSelection, newSelection: newSelection,
            added: added, removed: removed, event: event, cancel: false
        };
        this.grid.onRowSelectionChange.emit(args);
        if (args.cancel) {
            return;
        }
        this.selectRowsWithNoEvent(args.newSelection, true);
    };
    IgxGridSelectionService.prototype.getRowDataById = function (rowID) {
        if (!this.grid.primaryKey) {
            return rowID;
        }
        var rowIndex = this.getRowIDs(this.grid.gridAPI.get_all_data(true)).indexOf(rowID);
        return rowIndex < 0 ? {} : this.grid.gridAPI.get_all_data(true)[rowIndex];
    };
    IgxGridSelectionService.prototype.getRowIDs = function (data) {
        var _this = this;
        return this.grid.primaryKey && data.length ? data.map(function (rec) { return rec[_this.grid.primaryKey]; }) : data;
    };
    IgxGridSelectionService.prototype.clearHeaderCBState = function () {
        this.allRowsSelected = undefined;
    };
    /** Clear rowSelection and update checkbox state */
    IgxGridSelectionService.prototype.clearAllSelectedRows = function () {
        this.rowSelection.clear();
        this.clearHeaderCBState();
    };
    Object.defineProperty(IgxGridSelectionService.prototype, "allData", {
        /** Returns all data in the grid, with applied filtering and sorting and without deleted rows. */
        get: function () {
            var _this = this;
            var allData = this.isFilteringApplied() || this.grid.sortingExpressions.length ?
                this.grid.filteredSortedData : this.grid.gridAPI.get_all_data(true);
            return allData.filter(function (rData) { return !_this.isRowDeleted(_this.grid.gridAPI.get_row_id(rData)); });
        },
        enumerable: true,
        configurable: true
    });
    IgxGridSelectionService.prototype.areEqualCollections = function (first, second) {
        return first.length === second.length && new Set(first.concat(second)).size === first.length;
    };
    IgxGridSelectionService.prototype.isFilteringApplied = function () {
        var grid = this.grid;
        return !FilteringExpressionsTree.empty(grid.filteringExpressionsTree) ||
            !FilteringExpressionsTree.empty(grid.advancedFilteringExpressionsTree);
    };
    IgxGridSelectionService.prototype.isRowDeleted = function (rowID) {
        return this.grid.gridAPI.row_deleted_transaction(rowID);
    };
    /** Returns array of the selected columns fields. */
    IgxGridSelectionService.prototype.getSelectedColumns = function () {
        return this.columnSelection.size ? Array.from(this.columnSelection.keys()) : [];
    };
    IgxGridSelectionService.prototype.isColumnSelected = function (field) {
        return this.columnSelection.size > 0 && this.columnSelection.has(field);
    };
    /** Select the specified column and emit event. */
    IgxGridSelectionService.prototype.selectColumn = function (field, clearPrevSelection, selectColumnsRange, event) {
        var stateColumn = this.columnsState.field ? this.grid.getColumnByName(this.columnsState.field) : null;
        if (!event || !stateColumn || stateColumn.visibleIndex < 0 || !selectColumnsRange) {
            this.columnsState.field = field;
            this.columnsState.range = [];
            var newSelection = clearPrevSelection ? [field] : this.getSelectedColumns().indexOf(field) !== -1 ?
                this.getSelectedColumns() : __spread(this.getSelectedColumns(), [field]);
            var removed = clearPrevSelection ? this.getSelectedColumns().filter(function (colField) { return colField !== field; }) : [];
            var added = this.isColumnSelected(field) ? [] : [field];
            this.emitColumnSelectionEvent(newSelection, added, removed, event);
        }
        else if (selectColumnsRange) {
            this.selectColumnsRange(field, event);
        }
    };
    /** Select specified columns. And emit event. */
    IgxGridSelectionService.prototype.selectColumns = function (fields, clearPrevSelection, selectColumnsRange, event) {
        var _this = this;
        var columns = fields.map(function (f) { return _this.grid.getColumnByName(f); }).sort(function (a, b) { return a.visibleIndex - b.visibleIndex; });
        var stateColumn = this.columnsState.field ? this.grid.getColumnByName(this.columnsState.field) : null;
        if (!stateColumn || stateColumn.visibleIndex < 0 || !selectColumnsRange) {
            this.columnsState.field = columns[0] ? columns[0].field : null;
            this.columnsState.range = [];
            var added = fields.filter(function (colField) { return !_this.isColumnSelected(colField); });
            var removed = clearPrevSelection ? this.getSelectedColumns().filter(function (colField) { return fields.indexOf(colField) === -1; }) : [];
            var newSelection = clearPrevSelection ? fields : this.getSelectedColumns().concat(added);
            this.emitColumnSelectionEvent(newSelection, added, removed, event);
        }
        else {
            var filedStart = stateColumn.visibleIndex >
                columns[columns.length - 1].visibleIndex ? columns[0].field : columns[columns.length - 1].field;
            this.selectColumnsRange(filedStart, event);
        }
    };
    /** Select range from last clicked column to the current specified column. */
    IgxGridSelectionService.prototype.selectColumnsRange = function (field, event) {
        var _this = this;
        var currIndex = this.grid.getColumnByName(this.columnsState.field).visibleIndex;
        var newIndex = this.grid.columnToVisibleIndex(field);
        var columnsFields = this.grid.visibleColumns
            .filter(function (c) { return !c.columnGroup; })
            .sort(function (a, b) { return a.visibleIndex - b.visibleIndex; })
            .slice(Math.min(currIndex, newIndex), Math.max(currIndex, newIndex) + 1)
            .filter(function (col) { return col.selectable; }).map(function (col) { return col.field; });
        var removed = [];
        var oldAdded = [];
        var added = columnsFields.filter(function (colField) { return !_this.isColumnSelected(colField); });
        this.columnsState.range.forEach(function (f) {
            if (columnsFields.indexOf(f) === -1) {
                removed.push(f);
            }
            else {
                oldAdded.push(f);
            }
        });
        this.columnsState.range = columnsFields.filter(function (colField) { return !_this.isColumnSelected(colField) || oldAdded.indexOf(colField) > -1; });
        var newSelection = this.getSelectedColumns().concat(added).filter(function (c) { return removed.indexOf(c) === -1; });
        this.emitColumnSelectionEvent(newSelection, added, removed, event);
    };
    /** Select specified columns. No event is emitted. */
    IgxGridSelectionService.prototype.selectColumnsWithNoEvent = function (fields, clearPrevSelection) {
        var _this = this;
        if (clearPrevSelection) {
            this.columnSelection.clear();
        }
        fields.forEach(function (field) { _this.columnSelection.add(field); });
    };
    /** Deselect the specified column and emit event. */
    IgxGridSelectionService.prototype.deselectColumn = function (field, event) {
        this.initColumnsState();
        var newSelection = this.getSelectedColumns().filter(function (c) { return c !== field; });
        this.emitColumnSelectionEvent(newSelection, [], [field], event);
    };
    /** Deselect specified columns. No event is emitted. */
    IgxGridSelectionService.prototype.deselectColumnsWithNoEvent = function (fields) {
        var _this = this;
        fields.forEach(function (field) { return _this.columnSelection.delete(field); });
    };
    /** Deselect specified columns. And emit event. */
    IgxGridSelectionService.prototype.deselectColumns = function (fields, event) {
        var removed = this.getSelectedColumns().filter(function (colField) { return fields.indexOf(colField) > -1; });
        var newSelection = this.getSelectedColumns().filter(function (colField) { return fields.indexOf(colField) === -1; });
        this.emitColumnSelectionEvent(newSelection, [], removed, event);
    };
    IgxGridSelectionService.prototype.emitColumnSelectionEvent = function (newSelection, added, removed, event) {
        var currSelection = this.getSelectedColumns();
        if (this.areEqualCollections(currSelection, newSelection)) {
            return;
        }
        var args = {
            oldSelection: currSelection, newSelection: newSelection,
            added: added, removed: removed, event: event, cancel: false
        };
        this.grid.onColumnSelectionChange.emit(args);
        if (args.cancel) {
            return;
        }
        this.selectColumnsWithNoEvent(args.newSelection, true);
    };
    /** Clear columnSelection */
    IgxGridSelectionService.prototype.clearAllSelectedColumns = function () {
        this.columnSelection.clear();
    };
    IgxGridSelectionService.ctorParameters = function () { return [
        { type: NgZone }
    ]; };
    IgxGridSelectionService = __decorate([
        Injectable()
    ], IgxGridSelectionService);
    return IgxGridSelectionService;
}());
export { IgxGridSelectionService };
export function isChromium() {
    return (/Chrom|e?ium/g.test(navigator.userAgent) || /Google Inc/g.test(navigator.vendor)) && !/Edge/g.test(navigator.userAgent);
}
