// igniteui-angular/esm5/lib/grids/grid-mrl-navigation.service.js
import { __decorate, __extends, __read, __spread } from "tslib";
import { Injectable } from '@angular/core';
import { first } from 'rxjs/operators';
import { IgxGridNavigationService } from './grid-navigation.service';
import { HORIZONTAL_NAV_KEYS, HEADER_KEYS } from '../core/utils';
/** @hidden */
var IgxGridMRLNavigationService = /** @class */ (function (_super) {
    __extends(IgxGridMRLNavigationService, _super);
    function IgxGridMRLNavigationService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IgxGridMRLNavigationService.prototype.getNextPosition = function (rowIndex, colIndex, key, shift, ctrl, event) {
        if (!this.activeNode.layout) {
            this.activeNode.layout = this.layout(this.activeNode.column || 0);
        }
        switch (key) {
            case 'tab':
            case ' ':
            case 'spacebar':
            case 'space':
            case 'escape':
            case 'esc':
            case 'enter':
            case 'f2':
                _super.prototype.getNextPosition.call(this, rowIndex, colIndex, key, shift, ctrl, event);
                break;
            case 'end':
                rowIndex = ctrl ? this.findLastDataRowIndex() : this.activeNode.row;
                colIndex = ctrl ? this.lastColIndexPerMRLBlock(this.lastIndexPerRow) : this.lastIndexPerRow;
                break;
            case 'home':
                rowIndex = ctrl ? this.findFirstDataRowIndex() : this.activeNode.row;
                colIndex = ctrl ? 0 : this.firstIndexPerRow;
                break;
            case 'arrowleft':
            case 'left':
                colIndex = ctrl ? this.firstIndexPerRow : this.getNextHorizontalCellPosition(true).column;
                break;
            case 'arrowright':
            case 'right':
                colIndex = ctrl ? this.lastIndexPerRow : this.getNextHorizontalCellPosition().column;
                break;
            case 'arrowup':
            case 'up':
                var prevPos = this.getNextVerticalPosition(true);
                colIndex = ctrl ? this.activeNode.column : prevPos.column;
                rowIndex = ctrl ? this.findFirstDataRowIndex() : prevPos.row;
                break;
            case 'arrowdown':
            case 'down':
                var nextPos = this.getNextVerticalPosition();
                colIndex = ctrl ? this.activeNode.column : nextPos.column;
                rowIndex = ctrl ? this.findLastDataRowIndex() : nextPos.row;
                break;
            default:
                return;
        }
        var nextLayout = this.layout(colIndex);
        var newLayout = key.includes('up') || key.includes('down') ? { rowStart: nextLayout.rowStart } : { colStart: nextLayout.colStart };
        Object.assign(this.activeNode.layout, newLayout, { rowEnd: nextLayout.rowEnd });
        if (ctrl && (key === 'home' || key === 'end')) {
            this.activeNode.layout = nextLayout;
        }
        return { rowIndex: rowIndex, colIndex: colIndex };
    };
    IgxGridMRLNavigationService.prototype.isValidPosition = function (rowIndex, colIndex) {
        if (rowIndex < 0 || colIndex < 0 || this.grid.dataView.length - 1 < rowIndex ||
            Math.max.apply(Math, __spread(this.grid.visibleColumns.map(function (col) { return col.visibleIndex; }))) < colIndex ||
            (this.activeNode.column !== colIndex && !this.isDataRow(rowIndex, true))) {
            return false;
        }
        return true;
    };
    IgxGridMRLNavigationService.prototype.shouldPerformVerticalScroll = function (targetRowIndex, visibleColIndex) {
        if (!_super.prototype.shouldPerformVerticalScroll.call(this, targetRowIndex, visibleColIndex)) {
            return false;
        }
        if (!this.isDataRow(targetRowIndex)) {
            return _super.prototype.shouldPerformVerticalScroll.call(this, targetRowIndex, visibleColIndex);
        }
        var targetRow = _super.prototype.getRowElementByIndex.call(this, targetRowIndex);
        var containerHeight = this.grid.calcHeight ? Math.ceil(this.grid.calcHeight) : 0;
        var scrollPos = this.getVerticalScrollPositions(targetRowIndex, visibleColIndex);
        return (!targetRow || targetRow.offsetTop + scrollPos.topOffset < Math.abs(this.containerTopOffset)
            || containerHeight && containerHeight < scrollPos.rowBottom - Math.ceil(this.scrollTop));
    };
    IgxGridMRLNavigationService.prototype.isColumnFullyVisible = function (visibleColIndex) {
        var targetCol = this.grid.getColumnByVisibleIndex(visibleColIndex);
        if (this.isParentColumnFullyVisible(targetCol === null || targetCol === void 0 ? void 0 : targetCol.parent) || _super.prototype.isColumnPinned.call(this, visibleColIndex, this.forOfDir())) {
            return true;
        }
        var scrollPos = this.getChildColumnScrollPositions(visibleColIndex);
        var colWidth = scrollPos.rightScroll - scrollPos.leftScroll;
        if (this.displayContainerWidth < colWidth && this.displayContainerScrollLeft === scrollPos.leftScroll) {
            return true;
        }
        return this.displayContainerWidth >= scrollPos.rightScroll - this.displayContainerScrollLeft &&
            this.displayContainerScrollLeft <= scrollPos.leftScroll;
    };
    IgxGridMRLNavigationService.prototype.isParentColumnFullyVisible = function (parent) {
        if (!this.forOfDir().getScroll().clientWidth || (parent === null || parent === void 0 ? void 0 : parent.pinned)) {
            return true;
        }
        var index = this.forOfDir().igxForOf.indexOf(parent);
        return this.displayContainerWidth >= this.forOfDir().getColumnScrollLeft(index + 1) - this.displayContainerScrollLeft &&
            this.displayContainerScrollLeft <= this.forOfDir().getColumnScrollLeft(index);
    };
    IgxGridMRLNavigationService.prototype.getChildColumnScrollPositions = function (visibleColIndex) {
        var targetCol = this.grid.getColumnByVisibleIndex(visibleColIndex);
        var parentVIndex = this.forOfDir().igxForOf.indexOf(targetCol.parent);
        var leftScroll = this.forOfDir().getColumnScrollLeft(parentVIndex);
        var rightScroll = this.forOfDir().getColumnScrollLeft(parentVIndex + 1);
        targetCol.parent.children.forEach(function (c) {
            if (c.rowStart >= targetCol.rowStart && c.visibleIndex < targetCol.visibleIndex) {
                leftScroll += parseInt(c.width, 10);
            }
            if (c.rowStart <= targetCol.rowStart && c.visibleIndex > targetCol.visibleIndex) {
                rightScroll -= parseInt(c.width, 10);
            }
        });
        return { leftScroll: leftScroll, rightScroll: rightScroll };
    };
    IgxGridMRLNavigationService.prototype.getVerticalScrollPositions = function (rowIndex, visibleIndex) {
        var targetCol = this.grid.getColumnByVisibleIndex(visibleIndex);
        var rowSpan = targetCol.rowEnd && targetCol.rowEnd - targetCol.rowStart ? targetCol.rowEnd - targetCol.rowStart : 1;
        var topOffset = this.grid.defaultRowHeight * (targetCol.rowStart - 1);
        var rowTop = this.grid.verticalScrollContainer.sizesCache[rowIndex] + topOffset;
        return { topOffset: topOffset, rowTop: rowTop, rowBottom: rowTop + (this.grid.defaultRowHeight * rowSpan) };
    };
    IgxGridMRLNavigationService.prototype.performHorizontalScrollToCell = function (visibleColumnIndex, cb) {
        if (!this.shouldPerformHorizontalScroll(visibleColumnIndex)) {
            return;
        }
        var scrollPos = this.getChildColumnScrollPositions(visibleColumnIndex);
        var startScroll = scrollPos.rightScroll - this.displayContainerScrollLeft;
        var nextScroll = !(this.displayContainerScrollLeft <= scrollPos.leftScroll) && this.displayContainerWidth >= startScroll ?
            scrollPos.leftScroll : scrollPos.rightScroll - this.displayContainerWidth;
        this.forOfDir().getScroll().scrollLeft = nextScroll;
        this.grid.parentVirtDir.onChunkLoad
            .pipe(first())
            .subscribe(function () {
            if (cb) {
                cb();
            }
        });
    };
    IgxGridMRLNavigationService.prototype.performVerticalScrollToCell = function (rowIndex, visibleColIndex, cb) {
        var _a;
        var children = (_a = this.parentByChildIndex(visibleColIndex || 0)) === null || _a === void 0 ? void 0 : _a.children;
        if (!_super.prototype.isDataRow.call(this, rowIndex) || (children && children.length < 2)) {
            return _super.prototype.performVerticalScrollToCell.call(this, rowIndex, visibleColIndex, cb);
        }
        var containerHeight = this.grid.calcHeight ? Math.ceil(this.grid.calcHeight) : 0;
        var pos = this.getVerticalScrollPositions(rowIndex, visibleColIndex);
        var row = _super.prototype.getRowElementByIndex.call(this, rowIndex);
        if ((this.scrollTop > pos.rowTop) && (!row || row.offsetTop + pos.topOffset < Math.abs(this.containerTopOffset))) {
            pos.topOffset === 0 ? this.grid.verticalScrollContainer.scrollTo(rowIndex) :
                this.grid.verticalScrollContainer.scrollPosition = pos.rowTop;
        }
        else {
            this.grid.verticalScrollContainer.addScrollTop(Math.abs(pos.rowBottom - this.scrollTop - containerHeight));
        }
        this.grid.verticalScrollContainer.onChunkLoad
            .pipe(first()).subscribe(function () {
            if (cb) {
                cb();
            }
        });
    };
    IgxGridMRLNavigationService.prototype.getNextHorizontalCellPosition = function (previous) {
        var _this = this;
        if (previous === void 0) { previous = false; }
        var parent = this.parentByChildIndex(this.activeNode.column);
        if (!this.hasNextHorizontalPosition(previous, parent)) {
            return { row: this.activeNode.row, column: this.activeNode.column };
        }
        var columns = previous ? parent.children.filter(function (c) { return c.rowStart <= _this.activeNode.layout.rowStart; })
            .sort(function (a, b) { return b.visibleIndex - a.visibleIndex; }) : parent.children.filter(function (c) { return c.rowStart <= _this.activeNode.layout.rowStart; });
        var column = columns.find(function (col) { return previous ?
            col.visibleIndex < _this.activeNode.column && _this.rowEnd(col) > _this.activeNode.layout.rowStart :
            col.visibleIndex > _this.activeNode.column && col.colStart > _this.activeNode.layout.colStart; });
        if (!column || (previous && this.activeNode.layout.colStart === 1)) {
            var index_1 = previous ? parent.visibleIndex - 1 : parent.visibleIndex + 1;
            var children = this.grid.columnList.find(function (cols) { return cols.columnLayout && cols.visibleIndex === index_1; }).children;
            column = previous ? children.toArray().reverse().find(function (child) { return child.rowStart <= _this.activeNode.layout.rowStart; }) :
                children.find(function (child) { return _this.rowEnd(child) > _this.activeNode.layout.rowStart && child.colStart === 1; });
        }
        return { row: this.activeNode.row, column: column.visibleIndex };
    };
    IgxGridMRLNavigationService.prototype.getNextVerticalPosition = function (previous) {
        if (previous === void 0) { previous = false; }
        this.activeNode.column = this.activeNode.column || 0;
        if (!this.hasNextVerticalPosition(previous)) {
            return { row: this.activeNode.row, column: this.activeNode.column };
        }
        var currentRowStart = this.grid.getColumnByVisibleIndex(this.activeNode.column).rowStart;
        var nextBlock = !this.isDataRow(this.activeNode.row) ||
            (previous ? currentRowStart === 1 : currentRowStart === this.lastRowStartPerBlock());
        var nextRI = previous ? this.activeNode.row - 1 : this.activeNode.row + 1;
        if (nextBlock && !this.isDataRow(nextRI)) {
            return { row: nextRI, column: this.activeNode.column };
        }
        var children = this.parentByChildIndex(this.activeNode.column).children;
        var col = previous ? this.getPreviousRowIndex(children, nextBlock) : this.getNextRowIndex(children, nextBlock);
        return { row: nextBlock ? nextRI : this.activeNode.row, column: col.visibleIndex };
    };
    IgxGridMRLNavigationService.prototype.getNextRowIndex = function (children, next) {
        var _this = this;
        var rowStart = next ? 1 : this.rowEnd(this.grid.getColumnByVisibleIndex(this.activeNode.column));
        var col = children.filter(function (c) { return c.rowStart === rowStart; });
        return col.find(function (co) { return co.colStart === _this.activeNode.layout.colStart; }) ||
            col.sort(function (a, b) { return b.visibleIndex - a.visibleIndex; }).find(function (co) { return co.colStart <= _this.activeNode.layout.colStart; });
    };
    IgxGridMRLNavigationService.prototype.getPreviousRowIndex = function (children, prev) {
        var _this = this;
        var end = prev ? Math.max.apply(Math, __spread(children.map(function (c) { return _this.rowEnd(c); }))) :
            this.grid.getColumnByVisibleIndex(this.activeNode.column).rowStart;
        var col = children.filter(function (c) { return _this.rowEnd(c) === end; });
        return col.find(function (co) { return co.colStart === _this.activeNode.layout.colStart; }) ||
            col.sort(function (a, b) { return b.visibleIndex - a.visibleIndex; }).find(function (co) { return co.colStart <= _this.activeNode.layout.colStart; });
    };
    IgxGridMRLNavigationService.prototype.headerNavigation = function (event) {
        var key = event.key.toLowerCase();
        if (!HEADER_KEYS.has(key)) {
            return;
        }
        event.preventDefault();
        if (!this.activeNode.layout) {
            this.activeNode.layout = this.layout(this.activeNode.column || 0);
        }
        var alt = event.altKey;
        var ctrl = event.ctrlKey;
        this.performHeaderKeyCombination(this.grid.getColumnByVisibleIndex(this.activeNode.column), key, event.shiftKey, ctrl, alt, event);
        if (!ctrl && !alt && (key.includes('down') || key.includes('up'))) {
            var children = this.parentByChildIndex(this.activeNode.column).children;
            var col = key.includes('down') ? this.getNextRowIndex(children, false) : this.getPreviousRowIndex(children, false);
            if (!col) {
                return;
            }
            this.activeNode.column = col.visibleIndex;
            var newLayout = this.layout(this.activeNode.column);
            Object.assign(this.activeNode.layout, { rowStart: newLayout.rowStart, rowEnd: newLayout.rowEnd });
            return;
        }
        this.horizontalNav(event, key, -1);
    };
    IgxGridMRLNavigationService.prototype.horizontalNav = function (event, key, rowIndex) {
        var ctrl = event.ctrlKey;
        if (!HORIZONTAL_NAV_KEYS.has(key) || event.altKey) {
            return;
        }
        this.activeNode.row = rowIndex;
        if ((key.includes('left') || key === 'home') && this.activeNode.column > 0) {
            this.activeNode.column = ctrl || key === 'home' ? this.firstIndexPerRow : this.getNextHorizontalCellPosition(true).column;
        }
        if ((key.includes('right') || key === 'end') && this.activeNode.column !== this.lastIndexPerRow) {
            this.activeNode.column = ctrl || key === 'end' ? this.lastIndexPerRow : this.getNextHorizontalCellPosition().column;
        }
        var newLayout = this.layout(this.activeNode.column);
        Object.assign(this.activeNode.layout, { colStart: newLayout.colStart, rowEnd: newLayout.rowEnd });
        this.performHorizontalScrollToCell(this.activeNode.column);
    };
    Object.defineProperty(IgxGridMRLNavigationService.prototype, "lastIndexPerRow", {
        get: function () {
            var _this = this;
            var children = this.grid.visibleColumns.find(function (c) { return c.visibleIndex === _this.lastLayoutIndex && c.columnLayout; })
                .children.toArray().reverse();
            var column = children.find(function (co) { return co.rowStart === _this.activeNode.layout.rowStart; }) ||
                children.find(function (co) { return co.rowStart <= _this.activeNode.layout.rowStart; });
            return column.visibleIndex;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridMRLNavigationService.prototype, "firstIndexPerRow", {
        get: function () {
            var _this = this;
            var children = this.grid.visibleColumns.find(function (c) { return c.visibleIndex === 0 && c.columnLayout; }).children;
            var column = children.find(function (co) { return co.rowStart === _this.activeNode.layout.rowStart; }) ||
                children.find(function (co) { return co.rowStart <= _this.activeNode.layout.rowStart; });
            return column.visibleIndex;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridMRLNavigationService.prototype, "lastLayoutIndex", {
        get: function () {
            return Math.max.apply(Math, __spread(this.grid.visibleColumns.filter(function (c) { return c.columnLayout; }).map(function (col) { return col.visibleIndex; })));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridMRLNavigationService.prototype, "scrollTop", {
        get: function () {
            return Math.abs(this.grid.verticalScrollContainer.getScroll().scrollTop);
        },
        enumerable: true,
        configurable: true
    });
    IgxGridMRLNavigationService.prototype.lastColIndexPerMRLBlock = function (visibleIndex) {
        if (visibleIndex === void 0) { visibleIndex = this.activeNode.column; }
        return this.parentByChildIndex(visibleIndex).children.last.visibleIndex;
    };
    IgxGridMRLNavigationService.prototype.lastRowStartPerBlock = function (visibleIndex) {
        if (visibleIndex === void 0) { visibleIndex = this.activeNode.column; }
        return Math.max.apply(Math, __spread(this.parentByChildIndex(visibleIndex).children.map(function (c) { return c.rowStart; })));
    };
    IgxGridMRLNavigationService.prototype.rowEnd = function (column) {
        return column.rowEnd && column.rowEnd - column.rowStart ? column.rowStart + column.rowEnd - column.rowStart : column.rowStart + 1;
    };
    /**
     * @hidden
     * @internal
     */
    IgxGridMRLNavigationService.prototype.layout = function (visibleIndex) {
        var column = this.grid.getColumnByVisibleIndex(visibleIndex);
        return { colStart: column.colStart, rowStart: column.rowStart,
            colEnd: column.colEnd, rowEnd: column.rowEnd, columnVisibleIndex: column.visibleIndex };
    };
    IgxGridMRLNavigationService.prototype.parentByChildIndex = function (visibleIndex) {
        var _a;
        return (_a = this.grid.getColumnByVisibleIndex(visibleIndex)) === null || _a === void 0 ? void 0 : _a.parent;
    };
    IgxGridMRLNavigationService.prototype.hasNextHorizontalPosition = function (previous, parent) {
        if (previous === void 0) { previous = false; }
        if (previous && parent.visibleIndex === 0 && this.activeNode.layout.colStart === 1 ||
            !previous && parent.visibleIndex === this.lastLayoutIndex && this.activeNode.column === this.lastIndexPerRow) {
            return false;
        }
        return true;
    };
    IgxGridMRLNavigationService.prototype.hasNextVerticalPosition = function (prev) {
        if (prev === void 0) { prev = false; }
        if ((prev && this.activeNode.row === 0 && (!this.isDataRow(this.activeNode.row) || this.activeNode.layout.rowStart === 1)) ||
            (!prev && this.activeNode.row >= this.grid.dataView.length - 1 && this.activeNode.column === this.lastColIndexPerMRLBlock())) {
            return false;
        }
        return true;
    };
    IgxGridMRLNavigationService = __decorate([
        Injectable()
    ], IgxGridMRLNavigationService);
    return IgxGridMRLNavigationService;
}(IgxGridNavigationService));
export { IgxGridMRLNavigationService };
