// igniteui-angular/esm5/lib/grids/pinning/column-pinning.component.js
import { __decorate, __extends } from "tslib";
import { Component, ChangeDetectorRef } from '@angular/core';
import { ColumnChooserBaseDirective } from '../column-chooser-base';
import { IgxColumnPinningItemDirective } from './pinning.directive';
var IgxColumnPinningComponent = /** @class */ (function (_super) {
    __extends(IgxColumnPinningComponent, _super);
    function IgxColumnPinningComponent(cdr) {
        var _this = _super.call(this, cdr) || this;
        _this.cdr = cdr;
        return _this;
    }
    /**
     * @hidden
     */
    IgxColumnPinningComponent.prototype.createColumnItem = function (container, column) {
        if (column.level !== 0 || column.disablePinning) {
            return null;
        }
        var item = new IgxColumnPinningItemDirective();
        item.container = container;
        item.column = column;
        return item;
    };
    /**
     * @hidden
     */
    IgxColumnPinningComponent.prototype.checkboxValueChange = function (event, columnItem) {
        if (event.checked && !columnItem.pinnable) {
            event.checkbox.checked = false;
            return false;
        }
        columnItem.value = !columnItem.value;
    };
    Object.defineProperty(IgxColumnPinningComponent.prototype, "pinnableColumns", {
        /**
         * @hidden @internal
         */
        get: function () {
            return this.columnItems.filter(function (col) { return !col.pinningDisabled; });
        },
        enumerable: true,
        configurable: true
    });
    IgxColumnPinningComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    IgxColumnPinningComponent = __decorate([
        Component({
            preserveWhitespaces: false,
            selector: 'igx-column-pinning',
            template: "\n<div class=\"igx-column-hiding__header\">\n    <h4 class=\"igx-column-hiding__header-title\" *ngIf=\"title\">{{ title }}</h4>\n\n    <igx-input-group class=\"igx-column-hiding__header-input\" *ngIf=\"!disableFilter\">\n        <input igxInput\n            type=\"text\"\n            [(ngModel)]=\"filterCriteria\"\n            [placeholder]=\"filterColumnsPrompt\"\n            autocomplete=\"off\" />\n    </igx-input-group>\n</div>\n\n<div class=\"igx-column-hiding__columns\"\n    [style.max-height]=\"columnsAreaMaxHeight\">\n    <igx-checkbox igxColumnPinningItem\n        *ngFor=\"let columnItem of pinnableColumns\"\n        class=\"igx-column-hiding__columns-item\"\n        (change)=\"checkboxValueChange($event, columnItem)\"\n        [checked]=\"columnItem.value\">\n        {{ columnItem.name }}\n    </igx-checkbox>\n</div>\n"
        })
    ], IgxColumnPinningComponent);
    return IgxColumnPinningComponent;
}(ColumnChooserBaseDirective));
export { IgxColumnPinningComponent };
