// igniteui-angular/esm2015/lib/action-strip/grid-actions/grid-editing-actions.component.js
var IgxGridEditingActionsComponent_1;
import { __decorate } from "tslib";
import { Component, HostBinding } from '@angular/core';
import { IgxGridActionsBaseDirective } from './grid-actions-base.directive';
import { showMessage } from '../../core/deprecateDecorators';
let IgxGridEditingActionsComponent = IgxGridEditingActionsComponent_1 = class IgxGridEditingActionsComponent extends IgxGridActionsBaseDirective {
    constructor() {
        super(...arguments);
        /**
         * Host `class.igx-action-strip` binding.
         * @hidden
         * @internal
         */
        this.cssClass = 'igx-action-strip__editing-actions';
        this.isMessageShown = false;
    }
    /**
     * Enter row or cell edit mode depending the grid rowEditable option
     * @example
     * ```typescript
     * this.gridEditingActions.startEdit();
     * ```
     */
    startEdit(event) {
        if (event) {
            event.stopPropagation();
        }
        if (!this.isRow(this.strip.context)) {
            return;
        }
        const row = this.strip.context;
        const firstEditable = row.cells.filter(cell => cell.editable)[0];
        const grid = row.grid;
        if (!grid.hasEditableColumns) {
            this.isMessageShown = showMessage('The grid should be editable in order to use IgxGridEditingActionsComponent', this.isMessageShown);
            return;
        }
        // be sure row is in view
        if (grid.rowList.filter(r => r === row).length !== 0) {
            grid.crudService.begin(firstEditable);
        }
        this.strip.hide();
    }
    /**
     * Delete a row according to the context
     * @example
     * ```typescript
     * this.gridEditingActions.deleteRow();
     * ```
     */
    deleteRow(event) {
        if (event) {
            event.stopPropagation();
        }
        if (!this.isRow(this.strip.context)) {
            return;
        }
        const context = this.strip.context;
        const grid = context.grid;
        grid.deleteRow(context.rowID);
        this.strip.hide();
    }
    /**
     * Getter if the row is disabled
     * @hidden
     * @internal
     */
    get disabled() {
        if (!this.isRow(this.strip.context)) {
            return;
        }
        return this.strip.context.disabled;
    }
};
__decorate([
    HostBinding('class.igx-action-strip__editing-actions')
], IgxGridEditingActionsComponent.prototype, "cssClass", void 0);
IgxGridEditingActionsComponent = IgxGridEditingActionsComponent_1 = __decorate([
    Component({
        selector: 'igx-grid-editing-actions',
        template: "<ng-container *ngIf=\"isRowContext\">\n    <button igxButton=\"icon\" [disabled]=\"disabled\" igxRipple (click)=\"startEdit($event)\">\n        <igx-icon>edit</igx-icon>\n    </button>\n    <button igxButton=\"icon\" [disabled]=\"disabled\" class=\"igx-action-strip__delete\" igxRipple (click)=\"deleteRow($event)\">\n        <igx-icon >delete</igx-icon>\n    </button>\n</ng-container>\n",
        providers: [{ provide: IgxGridActionsBaseDirective, useExisting: IgxGridEditingActionsComponent_1 }]
    })
], IgxGridEditingActionsComponent);
export { IgxGridEditingActionsComponent };
