// igniteui-angular/esm5/lib/date-picker/calendar-container.component.js
import { __decorate } from "tslib";
import { Component, ViewChild, Input, Output, EventEmitter, HostListener, HostBinding } from '@angular/core';
import { InteractionMode } from '../core/enums';
/**
 * @hidden
 */
var IgxCalendarContainerComponent = /** @class */ (function () {
    function IgxCalendarContainerComponent() {
        this.mode = InteractionMode.Dialog;
        this.vertical = false;
        this.onClose = new EventEmitter();
        this.onTodaySelection = new EventEmitter();
        this.styleClass = 'igx-date-picker';
    }
    Object.defineProperty(IgxCalendarContainerComponent.prototype, "dropdownCSS", {
        get: function () {
            return this.mode === InteractionMode.DropDown;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarContainerComponent.prototype, "verticalCSS", {
        get: function () {
            return this.vertical && this.mode === InteractionMode.Dialog;
        },
        enumerable: true,
        configurable: true
    });
    IgxCalendarContainerComponent.prototype.onEscape = function (event) {
        event.preventDefault();
        this.onClose.emit();
    };
    Object.defineProperty(IgxCalendarContainerComponent.prototype, "isReadonly", {
        /**
         * Returns whether the date-picker is in readonly dialog mode.
         *
         * @hidden
         */
        get: function () {
            return this.mode === InteractionMode.Dialog;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Emits close event for the calendar.
     */
    IgxCalendarContainerComponent.prototype.closeCalendar = function () {
        this.onClose.emit();
    };
    /**
     * Emits today selection event for the calendar.
     */
    IgxCalendarContainerComponent.prototype.triggerTodaySelection = function () {
        this.onTodaySelection.emit();
    };
    __decorate([
        ViewChild('calendar', { static: true })
    ], IgxCalendarContainerComponent.prototype, "calendar", void 0);
    __decorate([
        Input()
    ], IgxCalendarContainerComponent.prototype, "mode", void 0);
    __decorate([
        Input()
    ], IgxCalendarContainerComponent.prototype, "vertical", void 0);
    __decorate([
        Input()
    ], IgxCalendarContainerComponent.prototype, "cancelButtonLabel", void 0);
    __decorate([
        Input()
    ], IgxCalendarContainerComponent.prototype, "todayButtonLabel", void 0);
    __decorate([
        Input()
    ], IgxCalendarContainerComponent.prototype, "datePickerActions", void 0);
    __decorate([
        Output()
    ], IgxCalendarContainerComponent.prototype, "onClose", void 0);
    __decorate([
        Output()
    ], IgxCalendarContainerComponent.prototype, "onTodaySelection", void 0);
    __decorate([
        HostBinding('class.igx-date-picker')
    ], IgxCalendarContainerComponent.prototype, "styleClass", void 0);
    __decorate([
        HostBinding('class.igx-date-picker--dropdown')
    ], IgxCalendarContainerComponent.prototype, "dropdownCSS", null);
    __decorate([
        HostBinding('class.igx-date-picker--vertical')
    ], IgxCalendarContainerComponent.prototype, "verticalCSS", null);
    __decorate([
        HostListener('keydown.esc', ['$event']),
        HostListener('keydown.alt.arrowup', ['$event'])
    ], IgxCalendarContainerComponent.prototype, "onEscape", null);
    IgxCalendarContainerComponent = __decorate([
        Component({
            selector: 'igx-calendar-container',
            template: "<ng-template #defaultDatePickerActions>\n    <div *ngIf=\"cancelButtonLabel || todayButtonLabel\" class=\"igx-date-picker__buttons\">\n        <button #closeButton *ngIf=\"cancelButtonLabel\" igxButton=\"flat\" igxRipple (click)=\"closeCalendar()\">\n            {{ cancelButtonLabel }}\n        </button>\n        <button #todayButton *ngIf=\"todayButtonLabel\" igxButton=\"flat\" igxRipple (click)=\"triggerTodaySelection()\">\n            {{ todayButtonLabel }}\n        </button>\n    </div>\n</ng-template>\n<igx-calendar #calendar></igx-calendar>\n<ng-container *ngTemplateOutlet=\"datePickerActions ? datePickerActions.template : defaultDatePickerActions\"></ng-container>\n",
            styles: [':host {display: block;}']
        })
    ], IgxCalendarContainerComponent);
    return IgxCalendarContainerComponent;
}());
export { IgxCalendarContainerComponent };
