// igniteui-angular/esm5/lib/date-picker/date-picker.utils.js
import { isIE } from '../core/utils';
import { DatePart } from '../directives/date-time-editor/date-time-editor.common';
var DATE_CHARS = ['h', 'H', 'm', 's', 'S', 't', 'T'];
var TIME_CHARS = ['d', 'D', 'M', 'y', 'Y'];
/** @hidden */
var DatePickerUtil = /** @class */ (function () {
    function DatePickerUtil() {
    }
    /**
     *  TODO: Unit tests for all public methods.
     */
    /**
     * Parse a Date value from masked string input based on determined date parts
     * @param inputData masked value to parse
     * @param dateTimeParts Date parts array for the mask
     */
    DatePickerUtil.parseValueFromMask = function (inputData, dateTimeParts, promptChar) {
        var _this = this;
        var parts = {};
        dateTimeParts.forEach(function (dp) {
            var value = parseInt(_this.getCleanVal(inputData, dp, promptChar), 10);
            if (!value) {
                value = dp.type === DatePart.Date || dp.type === DatePart.Month ? 1 : 0;
            }
            parts[dp.type] = value;
        });
        if (parts[DatePart.Month] < 1 || 12 < parts[DatePart.Month]) {
            return null;
        }
        // TODO: Century threshold
        if (parts[DatePart.Year] < 50) {
            parts[DatePart.Year] += 2000;
        }
        if (parts[DatePart.Date] > DatePickerUtil.daysInMonth(parts[DatePart.Year], parts[DatePart.Month])) {
            return null;
        }
        if (parts[DatePart.Hours] > 23 || parts[DatePart.Minutes] > 59 || parts[DatePart.Seconds] > 59) {
            return null;
        }
        return new Date(parts[DatePart.Year] || 2000, parts[DatePart.Month] - 1 || 0, parts[DatePart.Date] || 1, parts[DatePart.Hours] || 0, parts[DatePart.Minutes] || 0, parts[DatePart.Seconds] || 0);
    };
    /**
     * Parse the mask into date/time and literal parts
     */
    DatePickerUtil.parseDateTimeFormat = function (mask, locale) {
        if (locale === void 0) { locale = DatePickerUtil.DEFAULT_LOCALE; }
        var format = mask || DatePickerUtil.getDefaultInputFormat(locale);
        var dateTimeParts = [];
        var formatArray = Array.from(format);
        var currentPart = null;
        var position = 0;
        for (var i = 0; i < formatArray.length; i++, position++) {
            var type = DatePickerUtil.determineDatePart(formatArray[i]);
            if (currentPart) {
                if (currentPart.type === type) {
                    currentPart.format += formatArray[i];
                    if (i < formatArray.length - 1) {
                        continue;
                    }
                }
                DatePickerUtil.ensureLeadingZero(currentPart);
                currentPart.end = currentPart.start + currentPart.format.length;
                position = currentPart.end;
                dateTimeParts.push(currentPart);
            }
            currentPart = {
                start: position,
                end: position + formatArray[i].length,
                type: type,
                format: formatArray[i]
            };
        }
        return dateTimeParts;
    };
    DatePickerUtil.getDefaultInputFormat = function (locale) {
        if (!Intl || !Intl.DateTimeFormat || !Intl.DateTimeFormat.prototype.formatToParts) {
            // TODO: fallback with Intl.format for IE?
            return DatePickerUtil.SHORT_DATE_MASK;
        }
        var parts = DatePickerUtil.getDefaultLocaleMask(locale);
        parts.forEach(function (p) {
            if (p.type !== DatePart.Year && p.type !== DatePickerUtil.SEPARATOR) {
                p.formatType = "2-digit" /* TwoDigits */;
            }
        });
        return DatePickerUtil.getMask(parts);
    };
    DatePickerUtil.isDateOrTimeChar = function (char) {
        return DATE_CHARS.indexOf(char) !== -1 || TIME_CHARS.indexOf(char) !== -1;
    };
    DatePickerUtil.spinDate = function (delta, newDate, isSpinLoop) {
        var maxDate = DatePickerUtil.daysInMonth(newDate.getFullYear(), newDate.getMonth());
        var date = newDate.getDate() + delta;
        if (date > maxDate) {
            date = isSpinLoop ? date % maxDate : maxDate;
        }
        else if (date < 1) {
            date = isSpinLoop ? maxDate + (date % maxDate) : 1;
        }
        newDate.setDate(date);
    };
    DatePickerUtil.spinMonth = function (delta, newDate, isSpinLoop) {
        var maxDate = DatePickerUtil.daysInMonth(newDate.getFullYear(), newDate.getMonth() + delta);
        if (newDate.getDate() > maxDate) {
            newDate.setDate(maxDate);
        }
        var maxMonth = 11;
        var minMonth = 0;
        var month = newDate.getMonth() + delta;
        if (month > maxMonth) {
            month = isSpinLoop ? (month % maxMonth) - 1 : maxMonth;
        }
        else if (month < minMonth) {
            month = isSpinLoop ? maxMonth + (month % maxMonth) + 1 : minMonth;
        }
        newDate.setMonth(month);
    };
    DatePickerUtil.spinYear = function (delta, newDate) {
        var maxDate = DatePickerUtil.daysInMonth(newDate.getFullYear() + delta, newDate.getMonth());
        if (newDate.getDate() > maxDate) {
            // clip to max to avoid leap year change shifting the entire value
            newDate.setDate(maxDate);
        }
        newDate.setFullYear(newDate.getFullYear() + delta);
    };
    DatePickerUtil.spinHours = function (delta, newDate, isSpinLoop) {
        var maxHour = 23;
        var minHour = 0;
        var hours = newDate.getHours() + delta;
        if (hours > maxHour) {
            hours = isSpinLoop ? hours % maxHour - 1 : maxHour;
        }
        else if (hours < minHour) {
            hours = isSpinLoop ? maxHour + (hours % maxHour) + 1 : minHour;
        }
        newDate.setHours(hours);
    };
    DatePickerUtil.spinMinutes = function (delta, newDate, isSpinLoop) {
        var maxMinutes = 59;
        var minMinutes = 0;
        var minutes = newDate.getMinutes() + delta;
        if (minutes > maxMinutes) {
            minutes = isSpinLoop ? minutes % maxMinutes - 1 : maxMinutes;
        }
        else if (minutes < minMinutes) {
            minutes = isSpinLoop ? maxMinutes + (minutes % maxMinutes) + 1 : minMinutes;
        }
        newDate.setMinutes(minutes);
    };
    DatePickerUtil.spinSeconds = function (delta, newDate, isSpinLoop) {
        var maxSeconds = 59;
        var minSeconds = 0;
        var seconds = newDate.getSeconds() + delta;
        if (seconds > maxSeconds) {
            seconds = isSpinLoop ? seconds % maxSeconds - 1 : maxSeconds;
        }
        else if (seconds < minSeconds) {
            seconds = isSpinLoop ? maxSeconds + (seconds % maxSeconds) + 1 : minSeconds;
        }
        newDate.setSeconds(seconds);
    };
    DatePickerUtil.spinAmPm = function (newDate, currentDate, amPmFromMask) {
        switch (amPmFromMask) {
            case 'AM':
                newDate = new Date(newDate.setHours(newDate.getHours() + 12));
                break;
            case 'PM':
                newDate = new Date(newDate.setHours(newDate.getHours() - 12));
                break;
        }
        if (newDate.getDate() !== currentDate.getDate()) {
            return currentDate;
        }
        return newDate;
    };
    /**
     * Determines whether the provided value is greater than the provided max value.
     * @param includeTime set to false if you want to exclude time portion of the two dates
     * @param includeDate set to false if you want to exclude the date portion of the two dates
     * @returns true if provided value is greater than provided maxValue
     */
    DatePickerUtil.greaterThanMaxValue = function (value, maxValue, includeTime, includeDate) {
        if (includeTime === void 0) { includeTime = true; }
        if (includeDate === void 0) { includeDate = true; }
        // TODO: check if provided dates are valid dates and not Invalid Date
        // if maxValue is Invalid Date and value is valid date this will return:
        // - false if includeDate is true
        // - true if includeDate is false
        if (includeTime && includeDate) {
            return value.getTime() > maxValue.getTime();
        }
        var _value = new Date(value.getTime());
        var _maxValue = new Date(maxValue.getTime());
        if (!includeTime) {
            _value.setHours(0, 0, 0, 0);
            _maxValue.setHours(0, 0, 0, 0);
        }
        if (!includeDate) {
            _value.setFullYear(0, 0, 0);
            _maxValue.setFullYear(0, 0, 0);
        }
        return _value.getTime() > _maxValue.getTime();
    };
    /**
     * Determines whether the provided value is less than the provided min value.
     * @param includeTime set to false if you want to exclude time portion of the two dates
     * @param includeDate set to false if you want to exclude the date portion of the two dates
     * @returns true if provided value is less than provided minValue
     */
    DatePickerUtil.lessThanMinValue = function (value, minValue, includeTime, includeDate) {
        if (includeTime === void 0) { includeTime = true; }
        if (includeDate === void 0) { includeDate = true; }
        // TODO: check if provided dates are valid dates and not Invalid Date
        // if value is Invalid Date and minValue is valid date this will return:
        // - false if includeDate is true
        // - true if includeDate is false
        if (includeTime && includeDate) {
            return value.getTime() < minValue.getTime();
        }
        var _value = new Date(value.getTime());
        var _minValue = new Date(minValue.getTime());
        if (!includeTime) {
            _value.setHours(0, 0, 0, 0);
            _minValue.setHours(0, 0, 0, 0);
        }
        if (!includeDate) {
            _value.setFullYear(0, 0, 0);
            _minValue.setFullYear(0, 0, 0);
        }
        return _value.getTime() < _minValue.getTime();
    };
    DatePickerUtil.ensureLeadingZero = function (part) {
        switch (part.type) {
            case DatePart.Date:
            case DatePart.Month:
            case DatePart.Hours:
            case DatePart.Minutes:
            case DatePart.Seconds:
                if (part.format.length === 1) {
                    part.format = part.format.repeat(2);
                }
                break;
        }
    };
    DatePickerUtil.getCleanVal = function (inputData, datePart, promptChar) {
        return DatePickerUtil.trimEmptyPlaceholders(inputData.substring(datePart.start, datePart.end), promptChar);
    };
    DatePickerUtil.determineDatePart = function (char) {
        switch (char) {
            case 'd':
            case 'D':
                return DatePart.Date;
            case 'M':
                return DatePart.Month;
            case 'y':
            case 'Y':
                return DatePart.Year;
            case 'h':
            case 'H':
                return DatePart.Hours;
            case 'm':
                return DatePart.Minutes;
            case 's':
            case 'S':
                return DatePart.Seconds;
            case 't':
            case 'T':
                return DatePart.AmPm;
            default:
                return DatePart.Literal;
        }
    };
    /**
     * This method generates date parts structure based on editor mask and locale.
     * @param maskValue: string
     * @param locale: string
     * @returns array containing information about date parts - type, position, format
     */
    DatePickerUtil.parseDateFormat = function (maskValue, locale) {
        if (locale === void 0) { locale = DatePickerUtil.DEFAULT_LOCALE; }
        var dateStruct = [];
        if (maskValue === undefined && !isIE()) {
            dateStruct = DatePickerUtil.getDefaultLocaleMask(locale);
        }
        else {
            var mask = (maskValue) ? maskValue : DatePickerUtil.SHORT_DATE_MASK;
            var maskArray = Array.from(mask);
            var monthInitPosition = mask.indexOf("M" /* MonthChar */);
            var dayInitPosition = mask.indexOf("d" /* DayChar */);
            var yearInitPosition = mask.indexOf("y" /* YearChar */);
            if (yearInitPosition !== -1) {
                dateStruct.push({
                    type: "year" /* Year */,
                    initialPosition: yearInitPosition,
                    formatType: DatePickerUtil.getYearFormatType(mask)
                });
            }
            if (monthInitPosition !== -1) {
                dateStruct.push({
                    type: "month" /* Month */,
                    initialPosition: monthInitPosition,
                    formatType: DatePickerUtil.getMonthFormatType(mask)
                });
            }
            if (dayInitPosition !== -1) {
                dateStruct.push({
                    type: "day" /* Day */,
                    initialPosition: dayInitPosition,
                    formatType: DatePickerUtil.getDayFormatType(mask)
                });
            }
            for (var i = 0; i < maskArray.length; i++) {
                if (!DatePickerUtil.isDateChar(maskArray[i])) {
                    dateStruct.push({
                        type: DatePickerUtil.SEPARATOR,
                        initialPosition: i,
                        value: maskArray[i]
                    });
                }
            }
            dateStruct.sort(function (a, b) { return a.initialPosition - b.initialPosition; });
            DatePickerUtil.fillDatePartsPositions(dateStruct);
        }
        return dateStruct;
    };
    /**
     * This method generates input mask based on date parts.
     * @param dateStruct array
     * @returns input mask
     */
    DatePickerUtil.getInputMask = function (dateStruct) {
        var inputMask = [];
        for (var i = 0; i < dateStruct.length; i++) {
            if (dateStruct[i].type === DatePickerUtil.SEPARATOR) {
                inputMask.push(dateStruct[i].value);
            }
            else if (dateStruct[i].type === "day" /* Day */ || dateStruct[i].type === "month" /* Month */) {
                inputMask.push('00');
            }
            else if (dateStruct[i].type === "year" /* Year */) {
                switch (dateStruct[i].formatType) {
                    case "numeric" /* Numeric */: {
                        inputMask.push('0000');
                        break;
                    }
                    case "2-digit" /* TwoDigits */: {
                        inputMask.push('00');
                        break;
                    }
                }
            }
        }
        return inputMask.join('');
    };
    /**
     * This method generates editor mask.
     * @param dateStruct
     * @returns editor mask
     */
    DatePickerUtil.getMask = function (dateStruct) {
        var mask = [];
        for (var i = 0; i < dateStruct.length; i++) {
            switch (dateStruct[i].formatType) {
                case "numeric" /* Numeric */: {
                    if (dateStruct[i].type === "day" /* Day */) {
                        mask.push('d');
                    }
                    else if (dateStruct[i].type === "month" /* Month */) {
                        mask.push('M');
                    }
                    else {
                        mask.push('yyyy');
                    }
                    break;
                }
                case "2-digit" /* TwoDigits */: {
                    if (dateStruct[i].type === "day" /* Day */) {
                        mask.push('dd');
                    }
                    else if (dateStruct[i].type === "month" /* Month */) {
                        mask.push('MM');
                    }
                    else {
                        mask.push('yy');
                    }
                }
            }
            if (dateStruct[i].type === DatePickerUtil.SEPARATOR) {
                mask.push(dateStruct[i].value);
            }
        }
        return mask.join('');
    };
    /**
     * This method parses an input string base on date parts and returns a date and its validation state.
     * @param dateFormatParts
     * @param prevDateValue
     * @param inputValue
     * @returns object containing a date and its validation state
     */
    DatePickerUtil.parseDateArray = function (dateFormatParts, prevDateValue, inputValue) {
        var dayStr = DatePickerUtil.getDayValueFromInput(dateFormatParts, inputValue);
        var monthStr = DatePickerUtil.getMonthValueFromInput(dateFormatParts, inputValue);
        var yearStr = DatePickerUtil.getYearValueFromInput(dateFormatParts, inputValue);
        var yearFormat = DatePickerUtil.getDateFormatPart(dateFormatParts, "year" /* Year */).formatType;
        var day = (dayStr !== '') ? parseInt(dayStr, 10) : 1;
        var month = (monthStr !== '') ? parseInt(monthStr, 10) - 1 : 0;
        var year;
        if (yearStr === '') {
            year = (yearFormat === "2-digit" /* TwoDigits */) ? '00' : '2000';
        }
        else {
            year = yearStr;
        }
        var yearPrefix;
        if (prevDateValue) {
            var originalYear = prevDateValue.getFullYear().toString();
            if (originalYear.length === 4) {
                yearPrefix = originalYear.substring(0, 2);
            }
        }
        else {
            yearPrefix = '20';
        }
        var fullYear = (yearFormat === "2-digit" /* TwoDigits */) ? yearPrefix.concat(year) : year;
        if ((month < 0) || (month > 11) || (month === NaN)) {
            return { state: "invalid" /* Invalid */, value: inputValue };
        }
        if ((day < 1) || (day > DatePickerUtil.daysInMonth(fullYear, month)) || (day === NaN)) {
            return { state: "invalid" /* Invalid */, value: inputValue };
        }
        return { state: "valid" /* Valid */, date: new Date(fullYear, month, day) };
    };
    DatePickerUtil.maskToPromptChars = function (mask) {
        var result = mask.replace(/0|L/g, DatePickerUtil.PROMPT_CHAR);
        return result;
    };
    /**
     * This method replaces prompt chars with empty string.
     * @param value
     */
    DatePickerUtil.trimEmptyPlaceholders = function (value, promptChar) {
        var result = value.replace(new RegExp(promptChar || '_', 'g'), '');
        return result;
    };
    /**
     * This method is used for spinning date parts.
     * @param dateFormatParts
     * @param inputValue
     * @param position
     * @param delta
     * @param isSpinLoop
     * @return modified text input
     */
    DatePickerUtil.getModifiedDateInput = function (dateFormatParts, inputValue, position, delta, isSpinLoop) {
        var datePart = DatePickerUtil.getDatePartOnPosition(dateFormatParts, position);
        var datePartType = datePart.type;
        var datePartFormatType = datePart.formatType;
        var newValue;
        var datePartValue = DatePickerUtil.getDateValueFromInput(dateFormatParts, datePartType, inputValue);
        newValue = parseInt(datePartValue, 10);
        var maxValue, minValue;
        var minMax = DatePickerUtil.getMinMaxValue(dateFormatParts, datePart, inputValue);
        minValue = minMax.min;
        maxValue = minMax.max;
        if (isNaN(newValue)) {
            if (minValue === 'infinite') {
                newValue = 2000;
            }
            else {
                newValue = minValue;
            }
        }
        var tempValue = newValue;
        tempValue += delta;
        // Infinite loop for full years
        if (maxValue === 'infinite' && minValue === 'infinite') {
            newValue = tempValue;
        }
        if (isSpinLoop) {
            if (tempValue > maxValue) {
                tempValue = minValue;
            }
            if (tempValue < minValue) {
                tempValue = maxValue;
            }
            newValue = tempValue;
        }
        else {
            if (tempValue <= maxValue && tempValue >= minValue) {
                newValue = tempValue;
            }
        }
        var startIdx = datePart.position[0];
        var endIdx = datePart.position[1];
        var start = inputValue.slice(0, startIdx);
        var end = inputValue.slice(endIdx, inputValue.length);
        var changedPart;
        var prefix = DatePickerUtil.getNumericFormatPrefix(datePartFormatType);
        changedPart = (newValue < 10) ? "" + prefix + newValue : "" + newValue;
        return "" + start + changedPart + end;
    };
    /**
     * This method returns date input with prompt chars.
     * @param dateFormatParts
     * @param date
     * @param inputValue
     * @returns date input including prompt chars
     */
    DatePickerUtil.addPromptCharsEditMode = function (dateFormatParts, date, inputValue) {
        var dateArray = Array.from(inputValue);
        for (var i = 0; i < dateFormatParts.length; i++) {
            if (dateFormatParts[i].formatType === "numeric" /* Numeric */) {
                if ((dateFormatParts[i].type === "day" /* Day */ && date.getDate() < 10)
                    || (dateFormatParts[i].type === "month" /* Month */ && date.getMonth() + 1 < 10)) {
                    dateArray.splice(dateFormatParts[i].position[0], 0, DatePickerUtil.PROMPT_CHAR);
                    dateArray.join('');
                }
            }
        }
        return dateArray.join('');
    };
    /**
     * This method checks if date input is done.
     * @param dateFormatParts
     * @param input
     * @returns input completeness
     */
    DatePickerUtil.checkForCompleteDateInput = function (dateFormatParts, input) {
        var dayValue = DatePickerUtil.getDayValueFromInput(dateFormatParts, input);
        var monthValue = DatePickerUtil.getMonthValueFromInput(dateFormatParts, input);
        var yearValue = DatePickerUtil.getYearValueFromInput(dateFormatParts, input);
        var dayStr = DatePickerUtil.getDayValueFromInput(dateFormatParts, input, false);
        var monthStr = DatePickerUtil.getMonthValueFromInput(dateFormatParts, input, false);
        if (DatePickerUtil.isFullInput(dayValue, dayStr)
            && DatePickerUtil.isFullInput(monthValue, monthStr)
            && DatePickerUtil.isFullYearInput(dateFormatParts, yearValue)) {
            return 'complete';
        }
        else if (dayValue === '' && monthValue === '' && yearValue === '') {
            return 'empty';
        }
        else if (dayValue === '' || monthValue === '' || yearValue === '') {
            return 'partial';
        }
        return '';
    };
    DatePickerUtil.daysInMonth = function (fullYear, month) {
        return new Date(fullYear, month + 1, 0).getDate();
    };
    /**
     * Parse provided input to Date.
     * @param value input to parse
     * @returns Date if parse succeed or null
     */
    DatePickerUtil.parseDate = function (value) {
        if (typeof value === 'number') {
            return new Date(value);
        }
        // if value is Invalid Date we should return null
        if (this.isDate(value)) {
            return this.isValidDate(value) ? value : null;
        }
        return value ? new Date(Date.parse(value)) : null;
    };
    /**
     * Returns whether provided input is date
     * @param value input to check
     * @returns true if provided input is date
     */
    DatePickerUtil.isDate = function (value) {
        return Object.prototype.toString.call(value) === '[object Date]';
    };
    /**
     * Returns whether the input is valid date
     * @param value input to check
     * @returns true if provided input is a valid date
     */
    DatePickerUtil.isValidDate = function (value) {
        if (this.isDate(value)) {
            return !isNaN(value.getTime());
        }
        return false;
    };
    DatePickerUtil.getYearFormatType = function (format) {
        switch (format.match(new RegExp("y" /* YearChar */, 'g')).length) {
            case 1: {
                // y (2020)
                return "numeric" /* Numeric */;
            }
            case 4: {
                // yyyy (2020)
                return "numeric" /* Numeric */;
            }
            case 2: {
                // yy (20)
                return "2-digit" /* TwoDigits */;
            }
        }
    };
    DatePickerUtil.getMonthFormatType = function (format) {
        switch (format.match(new RegExp("M" /* MonthChar */, 'g')).length) {
            case 1: {
                // M (8)
                return "numeric" /* Numeric */;
            }
            case 2: {
                // MM (08)
                return "2-digit" /* TwoDigits */;
            }
        }
    };
    DatePickerUtil.getDayFormatType = function (format) {
        switch (format.match(new RegExp("d" /* DayChar */, 'g')).length) {
            case 1: {
                // d (6)
                return "numeric" /* Numeric */;
            }
            case 2: {
                // dd (06)
                return "2-digit" /* TwoDigits */;
            }
        }
    };
    DatePickerUtil.getDefaultLocaleMask = function (locale) {
        var dateStruct = [];
        var formatter = new Intl.DateTimeFormat(locale);
        var formatToParts = formatter.formatToParts(new Date());
        for (var i = 0; i < formatToParts.length; i++) {
            if (formatToParts[i].type === DatePickerUtil.SEPARATOR) {
                dateStruct.push({
                    type: DatePickerUtil.SEPARATOR,
                    value: formatToParts[i].value
                });
            }
            else {
                dateStruct.push({
                    type: formatToParts[i].type
                });
            }
        }
        var formatterOptions = formatter.resolvedOptions();
        for (var i = 0; i < dateStruct.length; i++) {
            switch (dateStruct[i].type) {
                case "day" /* Day */: {
                    dateStruct[i].formatType = formatterOptions.day;
                    break;
                }
                case "month" /* Month */: {
                    dateStruct[i].formatType = formatterOptions.month;
                    break;
                }
                case "year" /* Year */: {
                    dateStruct[i].formatType = formatterOptions.year;
                    break;
                }
            }
        }
        DatePickerUtil.fillDatePartsPositions(dateStruct);
        return dateStruct;
    };
    DatePickerUtil.isDateChar = function (char) {
        return (char === "y" /* YearChar */ || char === "M" /* MonthChar */ || char === "d" /* DayChar */);
    };
    DatePickerUtil.getNumericFormatPrefix = function (formatType) {
        switch (formatType) {
            case "2-digit" /* TwoDigits */: {
                return '0';
            }
            case "numeric" /* Numeric */: {
                return DatePickerUtil.PROMPT_CHAR;
            }
        }
    };
    DatePickerUtil.getMinMaxValue = function (dateFormatParts, datePart, inputValue) {
        var maxValue, minValue;
        switch (datePart.type) {
            case "month" /* Month */: {
                minValue = 1;
                maxValue = DatePickerUtil.NUMBER_OF_MONTHS;
                break;
            }
            case "day" /* Day */: {
                minValue = 1;
                maxValue = DatePickerUtil.daysInMonth(DatePickerUtil.getFullYearFromString(DatePickerUtil.getDateFormatPart(dateFormatParts, "year" /* Year */), inputValue), parseInt(DatePickerUtil.getMonthValueFromInput(dateFormatParts, inputValue), 10));
                break;
            }
            case "year" /* Year */: {
                if (datePart.formatType === "2-digit" /* TwoDigits */) {
                    minValue = 0;
                    maxValue = 99;
                }
                else {
                    // Infinite loop
                    minValue = 'infinite';
                    maxValue = 'infinite';
                }
                break;
            }
        }
        return { min: minValue, max: maxValue };
    };
    DatePickerUtil.getDateValueFromInput = function (dateFormatParts, type, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        var partPosition = DatePickerUtil.getDateFormatPart(dateFormatParts, type).position;
        var result = inputValue.substring(partPosition[0], partPosition[1]);
        return (trim) ? DatePickerUtil.trimEmptyPlaceholders(result) : result;
    };
    DatePickerUtil.getDayValueFromInput = function (dateFormatParts, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        return DatePickerUtil.getDateValueFromInput(dateFormatParts, "day" /* Day */, inputValue, trim);
    };
    DatePickerUtil.getMonthValueFromInput = function (dateFormatParts, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        return DatePickerUtil.getDateValueFromInput(dateFormatParts, "month" /* Month */, inputValue, trim);
    };
    DatePickerUtil.getYearValueFromInput = function (dateFormatParts, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        return DatePickerUtil.getDateValueFromInput(dateFormatParts, "year" /* Year */, inputValue, trim);
    };
    DatePickerUtil.getDateFormatPart = function (dateFormatParts, type) {
        var result = dateFormatParts.filter(function (datePart) { return (datePart.type === type); })[0];
        return result;
    };
    DatePickerUtil.isFullInput = function (value, input) {
        return (value !== '' && input.length === 2 && input.charAt(1) !== DatePickerUtil.PROMPT_CHAR);
    };
    DatePickerUtil.isFullYearInput = function (dateFormatParts, value) {
        switch (DatePickerUtil.getDateFormatPart(dateFormatParts, "year" /* Year */).formatType) {
            case "numeric" /* Numeric */: {
                return (value !== '' && value.length === 4);
            }
            case "2-digit" /* TwoDigits */: {
                return (value !== '' && value.length === 2);
            }
            default: {
                return false;
            }
        }
    };
    DatePickerUtil.getDatePartOnPosition = function (dateFormatParts, position) {
        var result = dateFormatParts.filter(function (element) {
            return element.position[0] <= position && position <= element.position[1] && element.type !== DatePickerUtil.SEPARATOR;
        })[0];
        return result;
    };
    DatePickerUtil.getFullYearFromString = function (yearPart, inputValue) {
        return parseInt(inputValue.substring(yearPart.position[0], yearPart.position[1]), 10);
    };
    DatePickerUtil.fillDatePartsPositions = function (dateArray) {
        var currentPos = 0;
        for (var i = 0; i < dateArray.length; i++) {
            // Day|Month part positions
            if (dateArray[i].type === "day" /* Day */ || dateArray[i].type === "month" /* Month */) {
                // Offset 2 positions for number
                dateArray[i].position = [currentPos, currentPos + 2];
                currentPos += 2;
            }
            else if (dateArray[i].type === "year" /* Year */) {
                // Year part positions
                switch (dateArray[i].formatType) {
                    case "numeric" /* Numeric */: {
                        // Offset 4 positions for full year
                        dateArray[i].position = [currentPos, currentPos + 4];
                        currentPos += 4;
                        break;
                    }
                    case "2-digit" /* TwoDigits */: {
                        // Offset 2 positions for short year
                        dateArray[i].position = [currentPos, currentPos + 2];
                        currentPos += 2;
                        break;
                    }
                }
            }
            else if (dateArray[i].type === DatePickerUtil.SEPARATOR) {
                // Separator positions
                dateArray[i].position = [currentPos, currentPos + 1];
                currentPos++;
            }
        }
    };
    DatePickerUtil.DEFAULT_INPUT_FORMAT = 'MM/dd/yyyy';
    // TODO: this is the def mask for the date-picker, should remove it during refactoring
    DatePickerUtil.SHORT_DATE_MASK = 'MM/dd/yy';
    DatePickerUtil.SEPARATOR = 'literal';
    DatePickerUtil.NUMBER_OF_MONTHS = 12;
    DatePickerUtil.PROMPT_CHAR = '_';
    DatePickerUtil.DEFAULT_LOCALE = 'en';
    return DatePickerUtil;
}());
export { DatePickerUtil };
