// igniteui-angular/esm5/lib/grids/moving/moving.drag.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, OnDestroy, Input, ElementRef, ViewContainerRef, NgZone, Renderer2, ChangeDetectorRef } from '@angular/core';
import { IgxDragDirective } from '../../directives/drag-drop/drag-drop.directive';
import { fromEvent } from 'rxjs';
import { IgxColumnMovingService } from './moving.service';
/**
 * @hidden
 * @internal
 */
var IgxColumnMovingDragDirective = /** @class */ (function (_super) {
    __extends(IgxColumnMovingDragDirective, _super);
    function IgxColumnMovingDragDirective(element, viewContainer, zone, renderer, cdr, cms) {
        var _this = _super.call(this, cdr, element, viewContainer, zone, renderer) || this;
        _this.element = element;
        _this.viewContainer = viewContainer;
        _this.zone = zone;
        _this.renderer = renderer;
        _this.cdr = cdr;
        _this.cms = cms;
        _this._ghostClass = 'igx-grid__drag-ghost-image';
        _this.ghostImgIconClass = 'igx-grid__drag-ghost-image-icon';
        _this.ghostImgIconGroupClass = 'igx-grid__drag-ghost-image-icon-group';
        _this.columnSelectedClass = 'igx-grid__th--selected';
        return _this;
    }
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "data", {
        set: function (val) {
            this._column = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "column", {
        get: function () {
            return this._column;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "draggable", {
        get: function () {
            return this.column && (this.column.movable || (this.column.groupable && !this.column.columnGroup));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "icon", {
        get: function () {
            return this.cms.icon;
        },
        enumerable: true,
        configurable: true
    });
    IgxColumnMovingDragDirective.prototype.ngOnDestroy = function () {
        this._unsubscribe();
    };
    IgxColumnMovingDragDirective.prototype.onEscape = function (event) {
        this.cms.cancelDrop = true;
        this.onPointerUp(event);
    };
    IgxColumnMovingDragDirective.prototype.onPointerDown = function (event) {
        var _this = this;
        if (!this.draggable || event.target.getAttribute('draggable') === 'false') {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        this._removeOnDestroy = false;
        this.cms.column = this.column;
        this.ghostClass = this._ghostClass;
        _super.prototype.onPointerDown.call(this, event);
        this.cms.isColumnMoving = true;
        this.column.grid.cdr.detectChanges();
        var args = {
            source: this.column
        };
        this.column.grid.onColumnMovingStart.emit(args);
        this.subscription$ = fromEvent(this.column.grid.document.defaultView, 'keydown').subscribe(function (ev) {
            if (ev.key === "Escape" /* ESCAPE */ || ev.key === "Esc" /* ESCAPE_IE */) {
                _this.onEscape(ev);
            }
        });
    };
    IgxColumnMovingDragDirective.prototype.onPointerMove = function (event) {
        event.preventDefault();
        _super.prototype.onPointerMove.call(this, event);
        if (this._dragStarted && this.ghostElement && !this.column.grid.draggedColumn) {
            this.column.grid.draggedColumn = this.column;
            this.column.grid.cdr.detectChanges();
        }
        if (this.cms.isColumnMoving) {
            var args = {
                source: this.column,
                cancel: false
            };
            this.column.grid.onColumnMoving.emit(args);
            if (args.cancel) {
                this.onEscape(event);
            }
        }
    };
    IgxColumnMovingDragDirective.prototype.onPointerUp = function (event) {
        var _this = this;
        // Run it explicitly inside the zone because sometimes onPointerUp executes after the code below.
        this.zone.run(function () {
            _super.prototype.onPointerUp.call(_this, event);
            _this.cms.isColumnMoving = false;
            _this.column.grid.draggedColumn = null;
            _this.column.grid.cdr.detectChanges();
        });
        this._unsubscribe();
    };
    IgxColumnMovingDragDirective.prototype.createGhost = function (pageX, pageY) {
        _super.prototype.createGhost.call(this, pageX, pageY);
        this.ghostElement.style.height = null;
        this.ghostElement.style.minWidth = null;
        this.ghostElement.style.flexBasis = null;
        this.ghostElement.style.position = null;
        this.renderer.removeClass(this.ghostElement, this.columnSelectedClass);
        var icon = document.createElement('i');
        var text = document.createTextNode('block');
        icon.appendChild(text);
        icon.classList.add('material-icons');
        this.cms.icon = icon;
        if (!this.column.columnGroup) {
            this.renderer.addClass(icon, this.ghostImgIconClass);
            this.ghostElement.insertBefore(icon, this.ghostElement.firstElementChild);
            this.ghostLeft = this._ghostStartX = pageX - ((this.ghostElement.getBoundingClientRect().width / 3) * 2);
            this.ghostTop = this._ghostStartY = pageY - ((this.ghostElement.getBoundingClientRect().height / 3) * 2);
        }
        else {
            this.ghostElement.insertBefore(icon, this.ghostElement.childNodes[0]);
            this.renderer.addClass(icon, this.ghostImgIconGroupClass);
            this.ghostElement.children[0].style.paddingLeft = '0px';
            this.ghostLeft = this._ghostStartX = pageX - ((this.ghostElement.getBoundingClientRect().width / 3) * 2);
            this.ghostTop = this._ghostStartY = pageY - ((this.ghostElement.getBoundingClientRect().height / 3) * 2);
        }
    };
    IgxColumnMovingDragDirective.prototype._unsubscribe = function () {
        if (this.subscription$) {
            this.subscription$.unsubscribe();
            this.subscription$ = null;
        }
    };
    IgxColumnMovingDragDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ViewContainerRef },
        { type: NgZone },
        { type: Renderer2 },
        { type: ChangeDetectorRef },
        { type: IgxColumnMovingService }
    ]; };
    __decorate([
        Input('igxColumnMovingDrag')
    ], IgxColumnMovingDragDirective.prototype, "data", null);
    IgxColumnMovingDragDirective = __decorate([
        Directive({
            selector: '[igxColumnMovingDrag]',
        })
    ], IgxColumnMovingDragDirective);
    return IgxColumnMovingDragDirective;
}(IgxDragDirective));
export { IgxColumnMovingDragDirective };
