// igniteui-angular/esm2015/lib/directives/for-of/virtual.helper.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, HostBinding, Input, ViewChild, ViewContainerRef, ChangeDetectorRef, OnDestroy, OnInit } from '@angular/core';
import { VirtualHelperBaseDirective } from './base.helper.component';
let VirtualHelperComponent = class VirtualHelperComponent extends VirtualHelperBaseDirective {
    constructor(elementRef, cdr) {
        super(elementRef, cdr);
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.cssClasses = 'igx-vhelper--vertical';
    }
    ngOnInit() {
        this.scrollWidth = this.getScrollWidth();
    }
    getScrollWidth() {
        const div = document.createElement('div');
        const style = div.style;
        style.width = '100px';
        style.height = '100px';
        style.position = 'absolute';
        style.top = '-10000px';
        style.top = '-10000px';
        style.overflow = 'scroll';
        document.body.appendChild(div);
        const scrollWidth = div.offsetWidth - div.clientWidth;
        document.body.removeChild(div);
        return scrollWidth ? scrollWidth + 1 : 0;
    }
};
VirtualHelperComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
__decorate([
    HostBinding('scrollTop')
], VirtualHelperComponent.prototype, "scrollTop", void 0);
__decorate([
    HostBinding('style.width.px')
], VirtualHelperComponent.prototype, "scrollWidth", void 0);
__decorate([
    ViewChild('container', { read: ViewContainerRef, static: true })
], VirtualHelperComponent.prototype, "_vcr", void 0);
__decorate([
    Input()
], VirtualHelperComponent.prototype, "itemsLength", void 0);
__decorate([
    HostBinding('class')
], VirtualHelperComponent.prototype, "cssClasses", void 0);
VirtualHelperComponent = __decorate([
    Component({
        selector: 'igx-virtual-helper',
        template: '<div #container class="igx-vhelper__placeholder-content" [style.height.px]="size"></div>'
    })
], VirtualHelperComponent);
export { VirtualHelperComponent };
