// igniteui-angular/esm5/lib/directives/for-of/virtual.helper.component.js
import { __decorate, __extends } from "tslib";
import { Component, ElementRef, HostBinding, Input, ViewChild, ViewContainerRef, ChangeDetectorRef, OnDestroy, OnInit } from '@angular/core';
import { VirtualHelperBaseDirective } from './base.helper.component';
var VirtualHelperComponent = /** @class */ (function (_super) {
    __extends(VirtualHelperComponent, _super);
    function VirtualHelperComponent(elementRef, cdr) {
        var _this = _super.call(this, elementRef, cdr) || this;
        _this.elementRef = elementRef;
        _this.cdr = cdr;
        _this.cssClasses = 'igx-vhelper--vertical';
        return _this;
    }
    VirtualHelperComponent.prototype.ngOnInit = function () {
        this.scrollWidth = this.getScrollWidth();
    };
    VirtualHelperComponent.prototype.getScrollWidth = function () {
        var div = document.createElement('div');
        var style = div.style;
        style.width = '100px';
        style.height = '100px';
        style.position = 'absolute';
        style.top = '-10000px';
        style.top = '-10000px';
        style.overflow = 'scroll';
        document.body.appendChild(div);
        var scrollWidth = div.offsetWidth - div.clientWidth;
        document.body.removeChild(div);
        return scrollWidth ? scrollWidth + 1 : 0;
    };
    VirtualHelperComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        HostBinding('scrollTop')
    ], VirtualHelperComponent.prototype, "scrollTop", void 0);
    __decorate([
        HostBinding('style.width.px')
    ], VirtualHelperComponent.prototype, "scrollWidth", void 0);
    __decorate([
        ViewChild('container', { read: ViewContainerRef, static: true })
    ], VirtualHelperComponent.prototype, "_vcr", void 0);
    __decorate([
        Input()
    ], VirtualHelperComponent.prototype, "itemsLength", void 0);
    __decorate([
        HostBinding('class')
    ], VirtualHelperComponent.prototype, "cssClasses", void 0);
    VirtualHelperComponent = __decorate([
        Component({
            selector: 'igx-virtual-helper',
            template: '<div #container class="igx-vhelper__placeholder-content" [style.height.px]="size"></div>'
        })
    ], VirtualHelperComponent);
    return VirtualHelperComponent;
}(VirtualHelperBaseDirective));
export { VirtualHelperComponent };
